include/boost/url/impl/any_allocator.hpp

```diff
@@ -0,0 +1,243 @@
+any_allocator(Allocator const& a) noexcept
+{
+    class impl;
+
+    using char_allocator_type = boost::allocator_rebind_t<Allocator, char>;
+
+    class impl : public base
+    {
+         char_allocator_type a_;
+
+         using aligned_1_t = boost::type_with_alignment<1>::type;
+         using aligned_2_t = boost::type_with_alignment<2>::type;
+         using aligned_4_t = boost::type_with_alignment<4>::type;
+         using aligned_8_t = boost::type_with_alignment<8>::type;
+         using aligned_16_t = boost::type_with_alignment<16>::type;
+
+         using aligned_1_allocator_type = boost::allocator_rebind_t<char_allocator_type, boost::type_with_alignment<1>::type>;
+         using aligned_2_allocator_type = boost::allocator_rebind_t<char_allocator_type, boost::type_with_alignment<2>::type>;
+         using aligned_4_allocator_type = boost::allocator_rebind_t<char_allocator_type, boost::type_with_alignment<4>::type>;
+         using aligned_8_allocator_type = boost::allocator_rebind_t<char_allocator_type, boost::type_with_alignment<8>::type>;
+         using aligned_16_allocator_type = boost::allocator_rebind_t<char_allocator_type, boost::type_with_alignment<16>::type>;
+         using aligned_max_allocator_type = boost::allocator_rebind_t<char_allocator_type, std::max_align_t>;
+
+    public:
+        ~impl() {}
+
+        explicit
+        impl(char_allocator_type const& a)
+        : base(1), a_(a)
+        {
+        }
+
+        static std::size_t ceil_div(std::size_t n, std::size_t d) {
+            return (n + d - 1) / d;
+        }
+
+        char*
+        allocate(std::size_t n, std::size_t size, std::size_t align) override
+        {
+            switch (align) {
+            case alignof(aligned_1_t):
+                if (size == 1) {
+                    return a_.allocate(n);
+                } else {
+                    return reinterpret_cast<char*>(aligned_1_allocator_type(a_).allocate(n * ceil_div(size, align)));
+                }
+            case alignof(aligned_2_t):
+                return reinterpret_cast<char*>(aligned_2_allocator_type(a_).allocate(n * ceil_div(size, align)));
+            case alignof(aligned_4_t):
+                return reinterpret_cast<char*>(aligned_4_allocator_type(a_).allocate(n * ceil_div(size, align)));
+            case alignof(aligned_8_t):
+                return reinterpret_cast<char*>(aligned_8_allocator_type(a_).allocate(n * ceil_div(size, align)));
+            case alignof(aligned_16_t):
+                return reinterpret_cast<char*>(aligned_16_allocator_type(a_).allocate(n * ceil_div(size, align)));
+            default:
+                return reinterpret_cast<char*>(aligned_max_allocator_type(a_).allocate(n * ceil_div(size, align)));
+            }
+        }
+
+        void
+        deallocate(char* p, std::size_t n, std::size_t size, std::size_t align) {
+            switch (align) {
+            case alignof(aligned_1_t): {
+                if (size == 1) {
```

Comment:
why are you checking size? why does this function take 4 parameters?

Reply:
It's the only way to calculate ceil_div, which is the way to calculate how much memory to allocate. Is the ceil_div expression wrong?